README.md
# NAME

Mojolicious::Plugin::Model::DB - It is an extension of the module [Mojolicious::Plugin::Model](https://metacpan.org/pod/Mojolicious::Plugin::Model)
for Mojolicious applications

# SYNOPSIS

Model Functions

    package MyApp::Model::Functions;
    use Mojo::Base 'MojoX::Model';

    sub trim {
        my ($self, $value) = @_;

        $value =~ s/^\s+|\s+$//g;

        return $value;
    }

    1;

Model DB Person

    package MyApp::Model::DB::Person;
    use Mojo::Base 'MojoX::Model';

    sub save {
        my ($self, $foo) = @_;

        return $self->mysql->db->insert(
            'foo',
            {
                foo => $foo
            }
        )->last_insert_id;
    }

    1;

Mojolicious::Lite application

    #!/usr/bin/env perl
    use Mojolicious::Lite;

    use lib 'lib';

    plugin 'Model::DB' => {mysql => 'mysql://user@/mydb'};

    any '/' => sub {
        my $c = shift;

        my $foo = $c->param('foo')
                ? $c->model('functions')->trim($c->param('foo')) # model functions
                : '';

        # model db Person
        my $id = $c->db('person')->save($foo);

        $c->render(text => $id);
    };

    app->start;

All available options

    #!/usr/bin/env perl
    use Mojolicious::Lite;

    plugin 'Model::DB' => {
        # Mojolicious::Plugin::Model::DB
        namespace    => 'DataBase',                # default is DB

        # databases options
        Pg           => 'postgresql://user@/mydb', # this will instantiate Mojo::Pg, in model get $self->pg,
        mysql        => 'mysql://user@/mydb',      # this will instantiate Mojo::mysql, in model get $self->mysql,
        SQLite       => 'sqlite:test.db',          # this will instantiate Mojo::SQLite, in model get $self->sqlite,
        Redis        => 'redis://localhost',       # this will instantiate Mojo::Redis, in model get $self->redis,
        
        # or using reference
        Pg           => Mojo::Pg->new('postgresql://postgres@/test'),       # in model get $self->pg,
        mysql        => Mojo::mysql->strict_mode('mysql://username@/test'), # in model get $self->mysql,
        SQLite       => Mojo::SQLite->new('sqlite:test.db'),                # in model get $self->sqlite,
        Redis        => Mojo::Redis->new("redis://localhost:6379/1"),       # in model get $self->redis,        

        # Mojolicious::Plugin::Model
        namespaces   => ['MyApp::Model', 'MyApp::CLI::Model'],
        base_classes => ['MyApp::Model'],
        default      => 'MyApp::Model::Pg',
        params       => {Pg => {uri => 'postgresql://user@/mydb'}}
    };

# DESCRIPTION

[Mojolicious::Plugin::Model::DB](https://metacpan.org/pod/Mojolicious::Plugin::Model::DB) It is an extension of the module Mojolicious::Plugin::Model,
the intention is to separate models of database from other models, using Mojolicious::Plugin::Model::DB you can continue using all functions of Mojolicious::Plugin::Model.
See more in [Mojolicious::Plugin::Model](https://metacpan.org/pod/Mojolicious::Plugin::Model)

# OPTIONS

## namespace

    # Mojolicious::Lite
    plugin 'Model::DB' => {namespace => 'DataBase'}; # It's will load from $moniker::DataBase

Namespace to load models from, defaults to `$moniker::DB`.

`$moniker::Model::DB` namespace is deprecated. It's will be removed in updates future.

## databases

#### Mojo::Pg

    # Mojolicious::Lite
    plugin 'Model::DB' => {Pg => 'postgresql://user@/mydb'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        return $self->pg->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;
    }

    1;

#### Mojo::mysql

    # Mojolicious::Lite
    plugin 'Model::DB' => {mysql => 'mysql://user@/mydb'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        return $self->mysql->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;
    }

    1;

#### Mojo::SQLite

    # Mojolicious::Lite
    plugin 'Model::DB' => {SQLite => 'sqlite:test.db'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        return $self->sqlite->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;
    }

    1;

#### Mojo::Redis

    # Mojolicious::Lite
    plugin 'Model::DB' => {Redis => 'redis://localhost'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $key) = @_;

        return $self->redis->db->get($key);
    }

    1;

#### Mojo::mysql and Mojo::Redis

    # Mojolicious::Lite
    plugin 'Model::DB' => {
        mysql => 'mysql://user@/mydb',
        Redis => 'redis://localhost'
    };

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        my $cache = $self->redis->db->get('foo:' . $id);
        return $cache if $cache;

        my $foo = $self->mysql->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;

        $self->redis->db->set('foo:' . $id, $foo);

        return $foo;
    }

    1;



## more options

see in [Mojolicious::Plugin::Model#OPTIONS](https://metacpan.org/pod/Mojolicious::Plugin::Model#OPTIONS)

# HELPERS

[Mojolicious::Plugin::Model::DB](https://metacpan.org/pod/Mojolicious::Plugin::Model::DB) implements the following helpers.

## db

    my $db = $c->db($name);

Load, create and cache a model object with given name. Default class for
model db `camelize($moniker)::Model::DB`. Return `undef` if model db not found.

## more helpers

see in [Mojolicious::Plugin::Model#HELPERS](https://metacpan.org/pod/Mojolicious::Plugin::Model#HELPERS)

# SEE ALSO

[Mojolicious](https://metacpan.org/pod/Mojolicious), [Mojolicious::Guides](https://metacpan.org/pod/Mojolicious::Guides),
[http://mojolicio.us](http://mojolicio.us), [Mojolicious::Plugin::Model](https://metacpan.org/pod/Mojolicious::Plugin::Model).

# AUTHOR

Lucas Tiago de Moraes `lucastiagodemoraes@gmail.com`

# COPYRIGHT AND LICENSE

This software is copyright (c) 2019 by Lucas Tiago de Moraes.

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.


Changes
1.06 2022-03-22T16:01:00Z
    - Added $moniker::DB namespace as default.
    - $moniker::Model::DB namespace is deprecated. It's will be removed in updates future.

1.05 2021-12-19T23:18:00Z
    - Added code to accept set database by reference

1.04 2021-07-13T07:50:00Z
    - Added die with the error if some database does not loader.
    - Added changes to work with latest version of Mojolicious::Plugin::Model.
    - Improved detection of default namespace like Mojolicious::Plugin::Model.

1.03 2021-06-27T00:50:00Z
    - Solved error in file test.

1.02 2021-06-27T00:26:00Z
    - Solved formatting of the pod.

1.01 2021-06-27T00:11:00Z
    - Added version minimal to Mojo::SQLite.
    - Added more information in doc.
    - Solved error in tests with Windows.

1.00 2021-06-25T21:11:00Z
    - Added Mojo::Pg, Mojo::mysql, Mojo::SQLite and Mojo::Redis in database options,
      auto instantiate modules and inject reference in Model::DB.
    - Solved big problem of camelize moniker.

0.10 2019-08-14T02:00:00Z
    - Added prerequisite minimum version 0.10 of module Mojolicious::Plugin::Model.

0.09 2019-08-07T09:40:00Z
    - Removed duplication of method _load_class_for_name.
    - Added module Storable to deep cloning variable $conf.
    - Remake 'namespaces' inside method register.

0.08 2019-08-04T00:30:00Z
    - Added missing prereq Mojo::Base as suggested by CPANTS. (manwar)

0.07 2019-08-04T00:30:00Z
    - Updated cpanfile

0.06 2019-08-04T00:15:00Z
    - Changed Moo to Class::Method::Modifiers

0.05 2019-08-04T00:01:00Z
    - Added module Moo like dependency.

0.04 2019-08-03T23:15:00Z
    - Fixed error in doc.

0.03 2019-08-03T23:03:00Z
    - Rename files MYMETA to META.

0.02 2019-08-03T10:14:40Z
    - Removed development trash.

0.01 2019-08-02T10:16:31Z
    - First release.


t/lib/Database/Model/DB/Demo.pm
package Database::Model::DB::Demo;
use Mojo::Base 'MojoX::Model';

sub find {
    my ($self, $id) = @_;

    return $self->sqlite->db->select(
        'demo',
        undef,
        {
            id => $id
        }
    )->hash;
}

1;


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Mojolicious::Plugin::Model::DB',
    ABSTRACT         => 'It is an extension of the module Mojolicious::Plugin::Model',
    AUTHOR           => 'Lucas Tiago de Moraes <lucastiagodemoraes@gmail.com>',
    VERSION_FROM     => 'lib/Mojolicious/Plugin/Model/DB.pm',
    LICENSE          => 'perl_5',
    MIN_PERL_VERSION => '5.010001',

    META_MERGE => {
        'meta-spec' => {
          version => 2,
        },

        resources => {
            homepage => 'https://github.com/lucas1/Mojolicious-Plugin-Model-DB',

            bugtracker => {
                web => 'https://github.com/lucas1/Mojolicious-Plugin-Model-DB/issues'
            },

            repository => {
                type => 'git',
                url  => 'git@github.com:lucas1/Mojolicious-Plugin-Model-DB.git',
                web  => 'https://github.com/lucas1/Mojolicious-Plugin-Model-DB'
            }
        }
    },

    PREREQ_PM => {
        'Storable' => 0,
        'Class::Method::Modifiers' => 0,
        'Mojolicious::Plugin::Model' => '0.11'
    },

    TEST_REQUIRES => {
        'Test::More' => 0,
        'Storable' => 0,
        'Class::Method::Modifiers' => 0,
        'Mojolicious::Plugin::Model' => '0.11',
        'Mojo::SQLite' => 0
    },

    test => {
        TESTS => 't/*.t'
    }
);


lib/Mojolicious/Plugin/Model/DB.pm
package Mojolicious::Plugin::Model::DB;
use Mojo::Base 'Mojolicious::Plugin::Model';
use Mojo::Util 'camelize';
use Mojo::Loader qw/load_class/;
use Storable qw/dclone/;
use Class::Method::Modifiers qw/after/;

our $VERSION = '1.06';

has 'databases' => sub {
    [qw/Pg mysql SQLite Redis/]
};

after register => sub {
    my ($plugin, $app, $conf) = @_;

    $conf = dclone $conf;

    # check if need camelize moniker
    my $moniker = camelize($app->moniker);
    $moniker    = $app->moniker unless -d $app->home . '/lib/' . $moniker;

    my $namespace  = $conf->{namespace}  // 'DB';
    my $namespaces = $conf->{namespaces} // [$moniker, $moniker . '::Model'];
    @{$conf->{namespaces}} = map $_ . "::$namespace", @$namespaces;
    my $databases = _load_class_databases($plugin, $conf);

    $app->helper(
        db => sub {
            my ($self, $name) = @_;
            $name //= $conf->{default};

            my $model;
            return $model if $model = $plugin->{models}{$name};

            my $class = Mojolicious::Plugin::Model::_load_class_for_name(
                $plugin, $app, $conf, $name, $moniker
            ) or return;
            
            # warning namespace deprecated
            if (!$conf->{namespaces} && $class =~ /::Model::${namespace}::/) {
                my ($moniker_current) = $class =~ /^([\w:]+)::Model::/;
                
                warn "${moniker_current}::Model::${namespace} namespace is deprecated. It will be removed on updates future!";
            }

            # define attr to database
            $class->attr($_) for keys %{$databases};

            my $params = $conf->{params}{$name};
            $model = $class->new(
                ref $params eq 'HASH' ? %$params : (),
                %$databases,
                app => $app
            );
            $plugin->{models}{$name} = $model;
            return $model;
        }
    );
};

sub _load_class_databases {
    my ($plugin, $conf) = @_;

    my $databases = {};

    for (@{$plugin->databases}) {
        if (defined $conf->{$_}) {
            my $class = 'Mojo::' . $_;
            
            my $ref;
            
            if (ref($conf->{$_}) eq $class) {
                $ref = $conf->{$_};
            } else {
                my $e     = load_class $class;
                die "Loading '$class' failed: $e" if $e;
                $ref = $class->new($conf->{$_});
            }
            
            $databases->{lc($_)} = $ref;
        }
    }

    return $databases;
}

1;

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Model::DB - It is an extension of the module L<Mojolicious::Plugin::Model> for Mojolicious applications.

=head1 SYNOPSIS

Model Functions

    package MyApp::Model::Functions;
    use Mojo::Base 'MojoX::Model';

    sub trim {
        my ($self, $value) = @_;

        $value =~ s/^\s+|\s+$//g;

        return $value;
    }

    1;

Model DB Person

    package MyApp::Model::DB::Person;
    use Mojo::Base 'MojoX::Model';

    sub save {
        my ($self, $foo) = @_;

        return $self->mysql->db->insert(
            'foo',
            {
                foo => $foo
            }
        )->last_insert_id;
    }

    1;

Mojolicious::Lite application

    #!/usr/bin/env perl
    use Mojolicious::Lite;

    use lib 'lib';

    plugin 'Model::DB' => {mysql => 'mysql://user@/mydb'};

    any '/' => sub {
        my $c = shift;

        my $foo = $c->param('foo')
                ? $c->model('functions')->trim($c->param('foo')) # model functions
                : '';

        # model db Person
        my $id = $c->db('person')->save($foo);

        $c->render(text => $id);
    };

    app->start;

All available options

    #!/usr/bin/env perl
    use Mojolicious::Lite;

    plugin 'Model::DB' => {
        # Mojolicious::Plugin::Model::DB
        namespace    => 'DataBase',                # default is DB

        # databases options
        Pg           => 'postgresql://user@/mydb', # this will instantiate Mojo::Pg, in model get $self->pg,
        mysql        => 'mysql://user@/mydb',      # this will instantiate Mojo::mysql, in model get $self->mysql,
        SQLite       => 'sqlite:test.db',          # this will instantiate Mojo::SQLite, in model get $self->sqlite,
        Redis        => 'redis://localhost',       # this will instantiate Mojo::Redis, in model get $self->redis,
        
        # or using reference
        Pg           => Mojo::Pg->new('postgresql://postgres@/test'),       # in model get $self->pg,
        mysql        => Mojo::mysql->strict_mode('mysql://username@/test'), # in model get $self->mysql,
        SQLite       => Mojo::SQLite->new('sqlite:test.db'),                # in model get $self->sqlite,
        Redis        => Mojo::Redis->new("redis://localhost:6379/1"),       # in model get $self->redis,

        # Mojolicious::Plugin::Model
        namespaces   => ['MyApp::Model', 'MyApp::CLI::Model'],
        base_classes => ['MyApp::Model'],
        default      => 'MyApp::Model::Pg',
        params       => {Pg => {uri => 'postgresql://user@/mydb'}}
    };

=head1 DESCRIPTION

Mojolicious::Plugin::Model::DB It is an extension of the module Mojolicious::Plugin::Model, the intention is to separate models of database from other models,
using Mojolicious::Plugin::Model::DB you can continue using all functions of Mojolicious::Plugin::Model. See more in L<Mojolicious::Plugin::Model>.

=head1 OPTIONS/  

=head2 namespace

    # Mojolicious::Lite
    plugin 'Model::DB' => {namespace => 'DataBase'}; # It's will load from $moniker::DataBase and $moniker::Model::DataBase

Namespace to load models from, defaults to C<$moniker::DB>.

C<$moniker::Model::DB> namespace is deprecated. It's will be removed in updates future.

=head2 databases

=head3 Mojo::Pg

    # Mojolicious::Lite
    plugin 'Model::DB' => {Pg => 'postgresql://user@/mydb'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        return $self->pg->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;
    }

    1;

=head3 Mojo::mysql

    # Mojolicious::Lite
    plugin 'Model::DB' => {mysql => 'mysql://user@/mydb'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        return $self->mysql->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;
    }

    1;

=head3 Mojo::SQLite

    # Mojolicious::Lite
    plugin 'Model::DB' => {SQLite => 'sqlite:test.db'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        return $self->sqlite->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;
    }

    1;

=head3 Mojo::Redis

    # Mojolicious::Lite
    plugin 'Model::DB' => {Redis => 'redis://localhost'};

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $key) = @_;

        return $self->redis->db->get($key);
    }

    1;

=head3 Mojo::mysql and Mojo::Redis

    # Mojolicious::Lite
    plugin 'Model::DB' => {
        mysql => 'mysql://user@/mydb',
        Redis => 'redis://localhost'
    };

    # Model::DB
    package MyApp::Model::DB::Foo;
    use Mojo::Base 'MojoX::Model';

    sub find {
        my ($self, $id) = @_;

        my $cache = $self->redis->db->get('foo:' . $id);
        return $cache if $cache;

        my $foo = $self->mysql->db->select(
            'foo',
            undef,
            {
                id => $id
            }
        )->hash;

        $self->redis->db->set('foo:' . $id, $foo);

        return $foo;
    }

    1;

=head2 more options

see in L<Mojolicious::Plugin::Model#OPTIONS>

=head1 HELPERS

L<Mojolicious::Plugin::Model::DB> implements the following helpers.

=head2 db

    my $db = $c->db($name);

Load, create and cache a model object with given name. Default class for
model db C<camelize($moniker)::Model::DB>. Return `undef` if model db not found.

=head2 more helpers

see in L<Mojolicious::Plugin::Model#HELPERS>

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<https://mojolicio.us>, L<Mojolicious::Plugin::Model>.

=head1 AUTHOR

Lucas Tiago de Moraes C<lucastiagodemoraes@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2019 by Lucas Tiago de Moraes.

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut
